Distinguish Imperial vs Metric Units based on the config

Abby-Wheelis opened the pull request:
Adapting chart generation and display to use the units requested in the config by the `display_config.use_imperial` field

This process will affect the chart generation notebooks and the html files, since some of the units are hard coded there, and we need to be fetching another parameter from the config file to be fed to the display and chart generation

This PR also serves as my first effort working on this side of the dashboard. 

Addresses #108

shankari reviewed (commented):
@Abby-Wheelis I am stopping the review here because there are some common themes that recur across the notebooks. Please let me know when you have addressed these and I will re-review.

The code looks fine overall, once you clean up these recurring patterns, and add the testing using `generate_plots.py`, I should be able to merge.

Abby-Wheelis commented on viz_scripts/generic_metrics.ipynb:
```diff
@@ -25,13 +25,14 @@
-    "year = 2020\n",
-    "month = 11\n",
+    "year = None\n",
+    "month = None\n",
```
Restored in [ca0f1e7](https://github.com/e-mission/em-public-dashboard/pull/114/commits/ca0f1e7d5c0493852ccb34b7984c71ac83fe01ff), I have tested by running the notebooks in the console, but I had so much trouble with the data maybe I messed something up? I will test again after I make these changes.

Abby-Wheelis commented on frontend/index.html:
```diff
@@ -391,7 +394,9 @@
-                                $('#metric').append(configuredResponse);
+																console.log("configuring units");
+																const unitConfigured = configuredResponse.replaceAll("${data.display_config.use_imperial}", dist_units);
+                                $('#metric').append(unitConfigured);
```
I went and updated all the formatting in that area of index.html, it was not that bad in VScode so I'm unsure why it was happening, but I think it looks better now

Abby-Wheelis commented on viz_scripts/generic_metrics.ipynb:
```diff
@@ -215,19 +216,33 @@
+    "\n",
+    "if use_imperial:\n",
+    "    dist_unit = \"miles\"\n",
+    "    dist_threshold = expanded_ct.distance_miles.quantile(0.8).round(1)\n",
```
here, it is different, if using imperial distance then the unit string in the labels is "miles" and the cutoff distance `dist_threshold` is conveyed in miles, for metric units both of those things are in kilometers

`cutoff` above is the raw distance (meters) that is what actually filters the data, the if/else just handles the two cases of labels

Abby-Wheelis replied:
I have now streamlined / deduplicated the if/else statements, using string variables as much as possible. There are still some that are similar, but switching them with a variable would be just as much code ... and typing that out I'm not sure that's a bad thing, if it meant all of the variables were assigned at once, it would be clear where things were getting switched ... I'm going to make one more pass

Abby-Wheelis commented on viz_scripts/generic_metrics.ipynb:
```diff
@@ -385,6 +440,9 @@
  "metadata": {
+  "interpreter": {
+   "hash": "72076fd04c06dcef1e530b48973026eda7036413ff5a3be73ace452fcce0cab6"
+  },
```
I think I've fixed all of these places now!

Abby-Wheelis commented:
I have now addressed unifying the if statements, removing extra cells, and reverting extraneous commits - I tried reverting the commit where most of that came in, but it broke several of the notebooks, so I removed them manually (I know this left the commit history a a mess but hopefully cleaned up the diff)

I still need to test everything again with `generate_plots.py`, but have run out of time today

Abby-Wheelis commented:
I ran through the testing and found several bugs which popped up when there was no data, no replaced modes, etc

I have been able to resolve these bugs so the notebooks run without error and the built in error handling can represent the errors smoothly. 

Testing was done by executing `PYTHONPATH=.. STUDY_CONFIG=cortezebikes python bin/generate_plots.py energy_calculations.ipynb default` for each of the notebooks in the `crontab`. After fixing the bugs I found, each runs smoothly. 

I then check the dashboard interface from it running locally, and found one more error, which I fixed in [36fdb88](https://github.com/e-mission/em-public-dashboard/pull/114/commits/36fdb8812a489f44cc93f69eacb775a55c76b904), and now all the charts are showing up as expected, except for Average trip length and Average trip length (sensed) which is known and under construction in #113.

shankari commented:
> so I removed them manually (I know this left the commit history a a mess but hopefully cleaned up the diff)

This is fine, I can just squash merge to unify the commit history

> Testing was done by executing PYTHONPATH=.. STUDY_CONFIG=cortezebikes python bin/generate_plots.py energy_calculations.ipynb default for each of the notebooks in the crontab. After fixing the bugs I found, each runs smoothly.

Just to clarify wrt https://github.com/e-mission/em-public-dashboard/pull/114#discussion_r1483239483
if you didn't revert the changes, the notebook would still run, but it would not override the month and year values. So we would end up with the same graphs for "All Data" and for each month. Can you confirm that this is not happening?

Can you confirm that this is not happening in your most recent set of tests?

Finally, I will review this now to give you feedback, but you need to resolve the merge conflicts from the previous merge of #113

Abby-Wheelis commented:
>Can you confirm that this is not happening in your most recent set of tests?

After resolving merge conflicts, I tested 1 more time and confirmed that charts are different month-to-month
<img width="762" alt="Screenshot 2024-02-10 at 2 01 04 PM" src="https://github.com/e-mission/em-public-dashboard/assets/54848919/9466fbf5-00bc-4614-b23b-99944618867b">

<img width="429" alt="Screenshot 2024-02-10 at 1 59 51 PM" src="https://github.com/e-mission/em-public-dashboard/assets/54848919/2fbd5de3-12de-4d06-855c-f4ba90a654ae">

shankari commented on viz_scripts/generic_timeseries.ipynb:
```diff
@@ -223,10 +237,7 @@
-    "    if (len(dynamic_labels)) > 0:\n",
-    "        plot_emission_per_week(emissions, 'kg', active_users)\n",
-    "    else:\n",
-    "        plot_emission_per_week(emissions, 'lb', active_users)\n",
+    "    plot_emission_per_week(emissions, weight_unit, active_users)\n",
     "except:\n",
```
Nice! that we were able to get rid of these repetitive `if/then` checks

Abby-Wheelis replied:
Your centralization suggestions have been extremely useful to centralization and reducing repeated code!

shankari reviewed (changes requested):
@Abby-Wheelis much better.

two changes needed:
- you can move the repeated section on the mode selection to scaffolding to reuse. This ensures that any fixes/changes can happen in one central location
- the index.html is still really messy. Can you revert the change and then turn off autoformatting in VS Code (if you have it enabled)? It may be that the original formatting is also bad, but I don't want to mix the formatting and code changes. If you can show that this is the case, I would accept a separate PR to fix the formatting, merge and then have the actual code changes show up here.

Abby-Wheelis commented on viz_scripts/generic_metrics.ipynb:
```diff
@@ -70,6 +71,25 @@
+    "label_units_lower = label_units.lower()\n",
+    "distance_col = \"distance_\" + short_label"
+   ]
+  },
```
That's a good plan! Done in [20c10f3](https://github.com/e-mission/em-public-dashboard/pull/114/commits/20c10f3289f3d8ff1d4ee61bbcd2c746feb72216)

Abby-Wheelis commented on viz_scripts/generic_metrics_sensed.ipynb:
```diff
@@ -166,30 +195,27 @@
    "source": [
-    "plot_title_no_quality=\"Miles for each primary mode\\n(inferred by OpenPATH from phone sensors)\"\n",
+    "plot_title_no_quality = label_units + \" for each primary mode\\n(inferred by OpenPATH from phone sensors)\"\n",
+    "plot_title = plot_title_no_quality + quality_text\n",
```
I moved it back! I think I glossed over the difference between `plot_title` and `plot_title_no_quality` thinking they both could go outside the try block - I also don't see how this could fail, but better to be safe (and match the convention in the rest of the file)[e89be8a](https://github.com/e-mission/em-public-dashboard/pull/114/commits/e89be8af2f431d13e4d1e474bd703cfbd2551891)

Abby-Wheelis commented:
Centralized to `scaffolding.py` and addressed the other few comments.

When it comes to index.html, I think that a reformat PR is the best route. Throughout the file the indentation style is inconsistent - going back and forth between 2 space and 4 space tabs in some places, extra indents that seem random (at least to me), I've submitted a single-commit PR #119 to take care of reformatting the whole file to have consistent formatting. I think this is a better change then trying to match the old way (which I did try to do and kept making it worse).

shankari commented:
@Abby-Wheelis everything looks fine except for `index.html` which still seems really messy?!

Abby-Wheelis commented:
@shankari I saw that when the merge for the formatting changes went through earlier and thought toggling the base would help but it did not, I'll resolve when I get the chance today, it should be the case that only my additions show in the diff for that file.

Abby-Wheelis commented:
Now the diff for `index.html` only shows changes that occurred as a part of this change - it was some combination of issues between GitHub and VScode
